File: 0039-combination-sum/0039-combination-sum.java
class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> res = new ArrayList<>();
        Arrays.sort(candidates);
        backtrack(res, 0, candidates, target, new ArrayList<>());
        return res;
    }

    public void backtrack(List<List<Integer>> res, int start, int[] nums, int remain, List<Integer> path){
        if(remain==0){
            res.add(new ArrayList<>(path));
            return;
        }

        for(int i=start; i<nums.length; i++){
            if(nums[i]>remain) break;

            path.add(nums[i]);
            backtrack(res, i, nums, remain-nums[i], path);
            path.remove(path.size()-1);
        }
    }
}

File: 0039-combination-sum/0039-combination-sum.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        res = []
        candidates.sort()
        def backtrack(cur, total, i):
            if total == target:
                res.append(cur.copy())
                return

            for j in range(i,len(candidates)):
                if total + candidates[j] > target:
                    break

                cur.append(candidates[j])
                backtrack(cur,total+candidates[j],j)
                cur.pop()


        backtrack([], 0, 0)
        return res